base/ca/src/org/dogtagpki/ct/CTEngine.java

```diff
@@ -0,0 +1,745 @@
+            tls_sct_ostream.write(intToFixedWidthBytes(tls_len, 2));
+            sct_ostream.writeTo(tls_sct_ostream);
+            byte[] tls_sct_bytes = tls_sct_ostream.toByteArray();
+
+            Extension ct_sct_ext = new Extension();
+            try (DerOutputStream out = new DerOutputStream()) {
+                out.putOctetString(tls_sct_bytes);
+                ct_sct_ext.setExtensionId(ct_sct_oid);
+                ct_sct_ext.setCritical(false);
+                ct_sct_ext.setExtensionValue(out.toByteArray());
+                logger.debug(method + " ct_sct_ext id = " +
+                    ct_sct_ext.getExtensionId().toString());
+                logger.debug(method + " CT extension constructed");
+            } catch (IOException e) {
+                logger.debug(method + e.toString());
+                return null;
+            } catch (Exception e) {
+                logger.debug(method + e.toString());
+                return null;
+            }
+
+            return ct_sct_ext;
+        } catch (Exception ex) {
+            logger.debug(method + ex.toString());
+            return null;
+        }
+    }
+
+    /**
+     * VerifySCT - conducts some verification of the SCT response returned
+     * from the CT log
+     *
+     * Two primary checks are intended to be performed in this method:
+     *
+     *   1. (works) verify that the log id in the CT response matches the CT log
+     *      signer public key hash;
+     *      - ToDo: consider putting the hash in CS.cfg to avoid run time
+     *        calculation
+     *   2. verify the signature in the CT against the SCT in
+     *        response
+     *      - concern: what if the extensions become out of order
+     *        during removal of the poison extension on the CT log server?
+     *        This could make it very difficult to check the signature,
+     *        since the CT response does not contain the tbsCert it signs.
+     *
+     * (Certificate Transparency)
+
+           digitally-signed struct {
+               Version sct_version;
+               SignatureType signature_type = certificate_timestamp; == 0 for ct
+               uint64 timestamp;
+               LogEntryType entry_type; ===> 1 for precert
+               select(entry_type) {
+                   case x509_entry: ASN.1Cert;
+                   case precert_entry: PreCert;
+               } signed_entry;
+              CtExtensions extensions;
+           };
+
+         struct { ==> 32 bit sha256 hash of issuer pub key + DER of precert
+           opaque issuer_key_hash[32];
+           TBSCertificate tbs_certificate;
+         } PreCert;
+    *
+    * @param response CT log server response
+    * @param tbsCert encoded TBSCert
+    * @param logPublicKey public key of log
+    * @return boolean true for verified; false for not verified
+    */
+    boolean verifySCT(CTResponse response, byte[] tbsCert, String logPublicKey, ICertificateAuthority ctCA) {
```

Comment:
Minor: This should be moved into `org.dogtagpki.ct.sct` as this is specific to the SCT. If we are going to reuse this same method for TimeStamps for other CT methods, this method needs to be renamed